### PR TITLE
Update to latest Styled Component types

### PR DESCRIPTION
**Pull request by jacobwgillespie**
Using `reakit` alongside the latest version of `@types/styled-components` results in an error since the TS types have changed for Styled Components and certain exports (like `StyledComponentClass`) no longer exist.  This PR updates the TypeScript types to their latest version.

It is not yet functional as upgrading introduces several TypeScript errors on build that will need to be fixed.

**What kind of change does this PR introduce?**

TypeScript types upgrade

**Does this PR introduce a breaking change?**

This is likely a breaking change since the TS generated types for Reakit would change, however this PR is opened against the `next` branch.

---

I'll keep exploring why the TS types are broken after this change - it seems to have something to do with `use` / `reuse` not picking up the props from `use`'d components.

The current errors are:

```

src/Avatar/__tests__/Avatar-test.tsx:7:13 - error TS2339: Property 'id' does not exist on type 'IntrinsicAttributes & Pick<{}, never> & Props<any> & { readonly use?: {} | {}[] | undefined; readonly useNext?: {} | {}[] | undefined; }'.

7     <Avatar id="test" aria-label="test" alt="test" />
              ~~~~~~~~~

... similar things for most tests ...
```

```

src/_utils/hoist.tsx:12:21 - error TS2352: Conversion of type 'ForwardRefExoticComponent<RefAttributes<{}>>' to type 'UseComponent<ComponentType<P>>' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
  Property 'uses' is missing in type 'ForwardRefExoticComponent<RefAttributes<{}>>'.

12   const Component = React.forwardRef((props, ref) => (
                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
13     <Comp {...props} elementRef={ref} />
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
14   )) as UseComponent<typeof Comp>;
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

src/_utils/hoist.tsx:13:22 - error TS2322: Type 'Ref<{}>' is not assignable to type '((instance: Element | null) => void) | RefObject<Element> | null | undefined'.
  Type 'RefObject<{}>' is not assignable to type '((instance: Element | null) => void) | RefObject<Element> | null | undefined'.
    Type 'RefObject<{}>' is not assignable to type 'RefObject<Element>'.
      Type '{}' is not assignable to type 'Element'.
        Property 'assignedSlot' is missing in type '{}'.

13     <Comp {...props} elementRef={ref} />
                        ~~~~~~~~~~

  src/_utils/hoist.tsx:5:3
    5   elementRef?: React.Ref<Element>;
        ~~~~~~~~~~
    The expected type comes from property 'elementRef' which is declared here on type 'IntrinsicAttributes & Props & { children?: ReactNode; }'
```

```

src/Provider/Provider.tsx:37:12 - error TS2322: Type '{ children: ReactNode; theme: T | ((theme: T) => T); }' is not assignable to type 'Readonly<ThemeProviderProps<any, any>>'.
  Types of property 'children' are incompatible.
    Type 'ReactNode' is not assignable to type 'string | number | ReactElement<any> | undefined'.
      Type 'null' is not assignable to type 'string | number | ReactElement<any> | undefined'.

37           <ThemeProvider theme={theme}>{children}</ThemeProvider>
              ~~~~~~~~~~~~~
```

My guess is the Provider fix is likely straightforward, and probably the test errors are related to why `hoist` is not passing type checks, but I haven't figured out if something needs to change in `reakit` or if a change in the types for `reuse` is necessary.

### CODE REVIEW

**Comment by jacobwgillespie** [comment]
I made a few changes, but there's something still wrong at the core of things.  I think it may have something to do with the type of `use` from Reuse, since components passed through `use` lose their Props definition.  I attempted to upgrade the React types in `reuse` today, but didn't get something that worked.  I opened https://github.com/diegohaz/reuse/pull/8 with my progress so far.

That PR does build, however it doesn't properly infer props when used with Reakit.  Hopefully you know more about what's going on with the TS types there than I do and can pinpoint what's missing. 🙂

**Comment by diegohaz** [comment]
Continuing from https://github.com/diegohaz/reuse/pull/8#issuecomment-443058035

I guess the error on the `as` prop happens because on Reakit we're defining `BoxProps` with `React.HTMLProps`:
https://github.com/reakit/reakit/blob/743458d6b37d4329b88a6f827f34d817f5decba8/packages/reakit/src/Box/Box.tsx#L19

Surprisingly, they define an `as` prop on that type. So they might be conflicting with the `as` prop coming from styled components. We can try to `Omit<>` it in `BoxProps`. That should fix the issue, but I'm not sure if that's an issue here or on styled-components typings (since styled components are supposed to accept all HTML attributes).

**Comment by jacobwgillespie** [comment]
I had to cast `hoist` through `unknown` - not sure if that's correct or if it's just masking some other problem.

File: packages/reakit/src/Box/Box.tsx [hunk 1]
```diff
@@ -14,9 +14,11 @@ import pickHTMLProps from "../_utils/pickHTMLProps";
 import styled from "../styled";
 import use from "../use";
 
+type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
```
Comment (diegohaz) [inline packages/reakit/src/Box/Box.tsx]: You can import `Omit` from `../_utils/types`

**Comment by diegohaz** [comment]
I published `reuse@1.3.0`. Could you upgrade the version here? Hopefully it's enough. :)

**Comment by jacobwgillespie** [comment]
Updated to reuse v1.3.0 - building now!

**Comment by diegohaz** [comment]
OK, now we have errors on `reakit-theme-default`. 

```
src/index.ts:83:14 - error TS4023: Exported variable 'Avatar' has or is using name 'FlattenSimpleInterpolation' from external module "/home/travis/build/reakit/reakit/node_modules/@types/styled-components/index" but cannot be named.
83 export const Avatar = css`
```

File: rollup.config.js [hunk 2]
```diff
@@ -37,6 +37,11 @@ const getPlugins = umd =>
         commonjs({
           include: /node_modules/,
           namedExports: {
+            "../../node_modules/react-is/index.js": [
```
Comment (wtgtybhertgeghgtwtg) [inline rollup.config.js:40]: Isn't this identical to the one below?

**Comment by diegohaz** [comment]
This is fixed on `next`.

**Pull request closed without merging**